### DISCUSSION TITLE
Common stutters from PrimeTween I am not sure what is causing this behaviour?

### DISCUSSION THREAD

**Discussion by NGrigorov** [KyryloKuzyk/PrimeTween#51] [FAQ] [answered]
![primetween](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/739cc4a0-358e-42ff-b571-99bec478783d)

I am proffiling my build and lets say roughly once every 5 to 10 seconds this spike happens.

Context:
My game is a tower defense game, sometimes a tower hits 5-10-15 enemies, and all of them do a PunchLocation for 1 sec duration and punch power is (Vector3.Up * 0.25f). No callback such as OnComplete is used.

Here is the code that uses this:

    public void Initialize(Vector3 location, float amount, Color col)
    {
        damangeInputText.text = amount.ToString("0");
        damangeInputText.color = col;
        transform.position = location;
        Tween.PunchLocalPosition(transform, Vector3.up * 0.25f, duration);
    }

**Comment by KyryloKuzyk** [comment]
Hmm, I can't reproduce this issue on my side in a stress test even with 1000 punches starting at the same time. Can you please provide more info that may help in debugging the issue?
Please also show the bottom part of the screen. Specifically the children of the onCycleComplete() method that have non-zero 'Time ms' column.

**Reply by NGrigorov** [reply]
Hey Kyrylo, thanks for the quick responce. I've been tinkering a little bit with URP settings since yesterday and I couldn't manage to reproduce it from the same call but I managed to produce another stutter similar to the previous one but from a shake callback:

![primetween2](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/f3e40eb5-d6b6-42dc-b6a9-5cc46107319a)
 I tried to expand on this image as much as possible, until the ms dissapeared.
 
 Additional images on the whole stack tree:
 
![primetween3](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/77bd6b8b-2bec-4de0-bea3-0e40c62df817)

![primetween4](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/6289df04-6a51-471a-ab1b-90d5cafba7b3)

Regarding additional info on the  punches I am not sure what I could say more since this is literally all I did: I was using the default tween but it produced a lot of gc for me when I did this (but no ms problems), so I just imported your package and replaced the default tween line with the version from your package. The GC problem was solved but i do get these stutters when lots of these are made in the same frame.

The trace for the creation of these is made like this:
1)
Enemies invoke a static method from the DmgPopUpManager when they get damaged. Each enemy has an instace of this script so this invoke might be 5 times or more in a frame. They send the damage they were just dealt, their position (this is not a transform injection, i save its position to a global variable whenever I adjust it on the transform, since lots of other scripts want this position often, i made sure to avoid constant transform call injections) and a color to label the dmg type.:

```        
private void HandleDamage(float healthDamage, float armorDamage, float shieldDamage)
        {
            if (this.currentShield > 0)
            {
                this.currentShield -= shieldDamage;
                DmgPopUpManager.Instance.OnDamageDealt?.Invoke(shieldDamage, enemyBehaviour.GetPosition(), Color.blue);
            }
            else if (this.currentArmor > 0)
            {
                this.currentArmor -= armorDamage;
                DmgPopUpManager.Instance.OnDamageDealt?.Invoke(armorDamage, enemyBehaviour.GetPosition(), Color.yellow);
            }
            else if (this.currentHealth > 0)
            {
                this.currentHealth -= healthDamage;
                DmgPopUpManager.Instance.OnDamageDealt?.Invoke(healthDamage, enemyBehaviour.GetPosition(), Color.green);
            }
        }
```
        
 From the DmgPopUpManager i have this void listening to the event:
 
 ```        
private void GenerateDamagePopUp(float amount, Vector3 location, Color col)
        {
            Profiler.BeginSample("GeneratePopUpDmg");
            if (amount <= 0.001)
                return;

            DmgPopUp dmgPopUp = ObjectPoolManager.SpawnObject(preafb.gameObject, Vector3.zero, Quaternion.identity, ObjectPoolManager.PoolType.DamageNum, ObjectPoolManager.ObjectType.DamangeNum).GetComponent<DmgPopUp>();
            dmgPopUp.Initialize(location, amount, col);
            Profiler.EndSample();
        }
```
        
Which just gets a dmgPopup prefab from teh pool, if none are available it will instantiate one instead. From the pool I get the component and initialize values for it.

On the DmgPopUp prefab this is the initialize:

```        
public void Initialize(Vector3 location, float amount, Color col)
        {
            damangeInputText.text = amount.ToString("0");
            damangeInputText.color = col;
            transform.position = location;
            Tween.PunchLocalPosition(transform, Vector3.up * 0.25f, duration).OnComplete(ShowText);
        }

        private void ShowText()
        {
            ObjectPoolManager.ReturnToPool(damangeInputText, ObjectPoolManager.ObjectType.DamangeNum);
        }
```
There are no other functions apart from those two. There are two variables to hold duration and the textmeshpro that adjust put from the inspector.
I added an OnComplete to return the dmgpopup to the pool after its done (it was an IEnumerator that yielded the duration before that generated GC and I didn't like that).

Maybe important note but, on the same prefab I have another script that rotates the prefab to look at the camera on LateUpdate(). But that will only happen if I rotate my camera otherwise a bool check will not permit that code to run. And this stutter happens even if I don't rotate the camera.

**Reply by NGrigorov** [reply]
I did another build without the Profiler.Sample and I have more information that might help you.

![primetween5](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/926b9b5b-23bc-4108-9570-7b0def16438f)
First I can confirm that its definitely not the Profiler that might be making this problem since here the primetween stuttered without the Sample in the function. But I did see very similar stutters in the proffiler, that were not caused by the tween but by the profiller.

Second I finally got some errors in the log file from the primetween:

![errorPrimeTween](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/de0c050d-b967-4d2b-8d1a-074b0e807b03)

Two logfiles, if you need them (I removed most of the debug.logs I use that are not related to make the file clear)
[priemtween.txt](https://github.com/KyryloKuzyk/PrimeTween/files/14352038/priemtween.txt)
[priemtween2.txt](https://github.com/KyryloKuzyk/PrimeTween/files/14352039/priemtween2.txt)


NOTE: I am **NOT** destroying the transform of the tween's target. I am **NOT** even disabling it, the object pooler is removing the text from the TextMeshPro variable and allowing it to be reused. in the scene again.

**Reply by NGrigorov** [reply]
Final Reply for now: I've tried to reproduce this behaviour in editor profiller but I can't.

**Comment by KyryloKuzyk** [comment]
@NGrigorov Thanks for the additional info!

Do you experience the same stutters without the Profiler? Profiler impacts performance by A LOT, especially with deep call stacks. And because two completely different call stacks show a similar stutter, I believe you're observing a Profiler's impact.

> Second I finally got some errors in the log file from the primetween. I am NOT destroying the transform of the tween's target. I am NOT even disabling it

The objects are destroyed when a scene is unloaded or changed. Can you please check if this is the case? Please also show the first error, typically it's the one that's the most relevant.

I would also appreciate a small reproducible example to isolate the issue from other dependencies. For example, this is a stress test I wrote trying to reproduce the issue, but with no luck. Tried it both on Android and macOS builds:
<details>
<summary>PrimeTweenExample.cs (click to expand)</summary>

```csharp
using PrimeTween;
using UnityEngine;

public class PrimeTweenExample : MonoBehaviour {
    [SerializeField] ShakeSettings shakeSettings = new ShakeSettings(new Vector3(0, 0.25f));
    [SerializeField] int count = 1000;
    Transform[] transforms;
    Tween[] tweens;

    void Awake() {
        transforms = new Transform[count];
        tweens = new Tween[count];
        for (int i = 0; i < count; i++) {
            transforms[i] = new GameObject().transform;
        }
    }

    void Update() {
        for (int i = 0; i < transforms.Length; i++) {
            if (!tweens[i].isAlive) {
                tweens[i] = Tween.PunchLocalPosition(transforms[i], shakeSettings); 
            }
        }
    }
}
```
</details>

**Comment by NGrigorov** [comment]
```Do you experience the same stutters without the Profiler? Profiler impacts performance by A LOT, especially with deep call stacks. And because two completely different call stacks show a similar stutter, I believe you're observing a Profiler's impact.```

To be honest I can't tell without the profiller :/. The stutter only happens for a single frame out of my target 60 and as of right now I can't tell, maybe once I get more stuff going on around it would be easier. But I am inclined to agree with you that its probably a deep profiller issue thats causing this. 
On my actual dev build without a profiller, I can see that my CPU spikes, whenever a hit happens that would use a lot of dmg pop ups but i am not sure if its because of the tween or because of the textmeshpro string assignment or because my pool instantiated a lot of popups. But the spike is 10% of my total so I am willing to "deal with it" for now. Also the spike quickly dissapears after it happens. (makes me belive the pool just instantiated some popups)

```The objects are destroyed when a scene is unloaded or changed. Can you please check if this is the case? Please also show the first error, typically it's the one that's the most relevant.```

I am 100% sure that this is why. My logger didn't show anything until I switched scenes, which means, while there were dmgPopups doing the OnComplete i switched scenes and the object got destroyed.

```I would also appreciate a small reproducible example to isolate the issue from other dependencies. For example, this is a stress test I wrote trying to reproduce the issue, but with no luck. Tried it both on Android and macOS builds:```

I'll try to make one later today that matches my setup but only the dmg popUps.

Thank you for the quick responces! @KyryloKuzyk

**Reply by NGrigorov** [reply]
@KyryloKuzyk I managed to succesfully replicate it on an empty project dev build with deep profiller attached with barebones implementation of my dmgPopUp behaviour.
It is however happening at roughly 200 popups, which is not even close to when it was happening in my case, but this could be the case because this project is not as heavy as mine so it has more memory to work with:

![primetweennew](https://github.com/KyryloKuzyk/PrimeTween/assets/15015079/5bb6c857-8f04-4658-a116-2970d478d779)
You can see in the image 4 spikes, each of them generates 50 dmg popups and at the 4th final one the stutter happened.

You can find the project on this repo if you want to take a look:
https://github.com/NGrigorov/PrimeTweenTest

**Notes:** The pool was already filled with more than 200 dmg popups so these are not instantiates.

**Steps to recreate:**
1. Build a dev version with deep proffiler attached
2. Start the build
3. Click the button 5-6 times to generate the instantiates so they do not interfere
4. Click 4 times consecutively not too far apart, make sure the previous click is still visible by the time you click again.
Result: Spike appears on the profiler

**Extra info:**
Unity Version: 2021.3.23f1

**Reply by KyryloKuzyk** [reply]
Huge thanks for the repro project! Will look at it soon. But the fact that these spikes happen at different unrelated moments leads me to think that it's just how Profiler works in the Deep Profile mode.
You don't see these half-a-second spikes without the Profiler enabled, right?

**Reply by KyryloKuzyk** [reply]
@NGrigorov Tried the repro project, and, unfortunately, I wasn't able to replicate the issue. The highest frame time I can get is 35 ms, nowhere near the 500 ms you're experiencing. I have an M1 macbook pro.

**Reply by NGrigorov** [reply]
@KyryloKuzyk Sorry to bother you one final time. I've pushed a "UnitTest" function to the same repo. Could you pull it again to test it one last time. If it doesn't work i'll leave it as is, and gonna assume its probably something from my machine or who knows what.

**Unitest:**
I've made a coroutine to generate the popUps automatically. Everytime I use this button I get the spike of 500ms.

Just make a dev build with a deep proffiler attached and click the button **ONCE**. If you don't get the same behaviour then we can leave it as is here :D

**Reply by KyryloKuzyk** [reply, marked as answer]
@NGrigorov Nope, still nothing. Maybe I use the latest version of Unity 2021.3.34f instead of yours 2021.3.23f. But maybe this issue doesn't exist on M1 chip. 

In any case, Profiler is an approximate tool designed to look for bottlenecks and give an approximate picture. But it's not the perfect tool to measure actual real-world performance. This is why I use [Performance testing API](https://github.com/KyryloKuzyk/PrimeTween/discussions/10) in my tests.

**Reply by NGrigorov** [reply]
Thanks for the update and the help. Lets mark it as solved!
